src/fft.jl

```diff
@@ -1,63 +1,121 @@
 represented in the spherical basis sets, `supersampling` should be at least `2`.
 """
-function determine_grid_size(lattice::AbstractMatrix{T}, Ecut; supersampling=2, tol=1e-8,
-                             ensure_smallprimes=true) where T
-    cutoff_Gsq = 2 * supersampling^2 * Ecut
-    Gmax = [norm(lattice[:, i]) / 2T(π) * sqrt(cutoff_Gsq) for i in 1:3]
+function determine_fft_size(lattice::AbstractMatrix{T}, Ecut; supersampling=2, tol=1e-8,
+                            ensure_smallprimes=true) where T
+    Gmax = supersampling * sqrt(2 * Ecut)
+    Glims = bounding_rectangle(lattice, Gmax; tol=tol)
 
-    # Round up, unless exactly zero (in which case keep it zero in
-    # order to just have one G vector for 1D or 2D systems)
-    for i = 1:3
-        if Gmax[i] != 0
-            Gmax[i] = ceil.(Int, Gmax[i] .- tol)
+    fft_size = Vec3(2 .* Glims .+ 1)
+    # Optimize FFT grid size: Make sure the number factorises in small primes only
+    if ensure_smallprimes
+        fft_size = nextprod.(Ref(_smallprimes()), fft_size)
+    end
+    fft_size
+end
+function determine_fft_size(model::Model, Ecut; kwargs...)
+    determine_fft_size(model.lattice, Ecut; kwargs...)
+end
+function lattice_diameter(latt)
+    diam = zero(eltype(latt))
+    # brute force search
+    for i in (-1, 0, 1)
+        for j in (-1, 0, 1)
+            for k in (-1, 0, 1)
+                diam = max(diam, norm(latt*[i; j; k]))
+            end
         end
     end
+    diam
+end
 
+## This uses a more precise and slower algorithm than the one above,
+## simply enumerating all G vectors and seeing where their difference
+## is. It needs the kpoints to do so.
+function determine_fft_size_precise(lattice::AbstractMatrix{T}, Ecut, kpoints;
+                                    supersampling=2, ensure_smallprimes=true) where T
+    recip_lattice = 2T(π)*pinv(lattice')  # pinv in case one of the dimension is trivial
+    recip_diameter = lattice_diameter(recip_lattice)
+    Glims = [0, 0, 0]
+    # get the bounding rectangle the naive way
+    for kpt in kpoints
+        for G in G_vectors(kpt)
+            if norm(recip_lattice * (G+kpt.coordinate)) ≤ sqrt(2Ecut) - recip_diameter
+                # each of the 8 neighbors (in ∞-norm) also belongs to the grid
+                # so we can safely skip the search knowing at least one of them
+                # will have larger |G-Gp|.
+                # Savings with this trick are surprisingly small : for silicon, 50% at Ecut 30, 70% at Ecut 100
+                continue
+            end
+            for Gp in G_vectors(kpt)
+                for i = 1:3
+                    @inbounds Glims[i] = max(Glims[i], abs(G[i] - Gp[i]))
+                end
+            end
+        end
+    end
+    if supersampling != 2
+        # no guarantees there, we just do our best to satisfy the
+        # target supersampling ratio
+        Glims = round.(Int, supersampling ./ 2 .* Glims)
```

Comment:
Just took me a bit to understand this mechanics. Could you add a comment why you compare supersampling to `2` and divide by `2` (in the sense that the double loop exactly covers the conventional case supersampling of 2, so if it's not that you just scale).

Reply:
added a comment above